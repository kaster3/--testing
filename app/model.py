from dataclasses import dataclass
from enum import Enum


class BookStatus(Enum):
    """
    Клас представляющий статус книг (в наличии, выдана), для более удобного взаимодействия
    и уменьшения шанса на ошибку
    """
    IN_STOCK = "В наличии"
    CHECKED_OUT = "Выдана"

    @staticmethod
    def from_value(value: str) -> "BookStatus":
        """
        Метод для получения статуса книги по его значению, для того чтобы json корректно работал с Enum
        """
        for status in BookStatus:
            if status.value == value:
                return status
        return BookStatus.IN_STOCK


@dataclass
class Book:
    """Основная бизнес сущность с которой будут выполняться все CRUD операции"""
    id: int
    title: str
    author: str
    year: int
    status: BookStatus = BookStatus.IN_STOCK


    def __str__(self):
        """
        Метод для вывода информации о книге в удобном виде
        """
        return (
            f"Книга №{self.id}\nНазвание: {self.title}\nАвтор:  {self.author}\n"
            f"Год издания: {self.year}\nСтатус выдачи: {self.status.value}\n"
        )


    def to_dict(self):
        """
        Метод для преобразования экземпляра книги в словарь так как json не умеет работать с классом Enum
        """
        return {
            'id': self.id,
            'title': self.title,
            'author': self.author,
            'year': self.year,
            'status': self.status.value
        }


    def __hash__(self):
        return hash((self.id, self.title, self.author, self.year, self.status))

    def __eq__(self, other):
        if not isinstance(other, Book):
            return NotImplemented

        return (self.id, self.title, self.author, self.year, self.status) == \
            (other.id, other.title, other.author, other.year, other.status)