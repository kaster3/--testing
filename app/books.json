[
    {
        "id": 1,
        "title": "Traveling",
        "author": "Shara",
        "year": 1999,
        "status": "В наличии"
    },
    {
        "id": 2,
        "title": "Traveling",
        "author": "John Smith",
        "year": 1999,
        "status": "В наличии"
    },
    {
        "id": 3,
        "title": "Terminator",
        "author": "Albec Ruso",
        "year": 1979,
        "status": "В наличии"
    }
]