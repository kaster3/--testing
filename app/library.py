from .storage import Storage
from .model import Book, BookStatus


class LibraryCRUD:
    """
    Класс LibraryCRUD предназначен для прямого выполнения CRUD операций, а также работает
    с хранилищем книг, сохраняя туда данные и вытаскивая

    Args:
        self.storage: Класс хранилище
        self.books: Список книг, загруженных из json файла
    """
    def __init__(self, storage: Storage) -> None:
        self.storage: Storage = storage
        self.books: list[Book | None] = self.storage.load_books()


    def get_book_by_id(self, book_id: int | None) -> Book | None:
        """
        Метод для получения книги по ее id
        """
        for book in self.books:
            if book.id == book_id:
                return book


    def add_book(self, title: str, author: str, year: int) -> None:
        """
        Метод для добавления новой книги и ее сохранения на основе полученных параметров
        с автоматической генерации уникального id книги
        """
        # Генерация уникального id
        book_id: int = self.books[-1].id + 1 if self.books else 1
        new_book: Book = Book(book_id, title, author, year)
        self.books.append(new_book)
        self.storage.save_books(books=self.books)


    def remove_book(self, book: Book) -> None:
        """
        Метод удаления книги предварительно, которая была найдена
        """
        self.books.remove(book)
        self.storage.save_books(books=self.books)


    def get_books(
            self,
            title: str | None = None,
            author: str | None = None,
            year: int | None = None,
    ) -> set[Book] | list[Book | None]:
        """
        Метод получения, неповторяющихся, книг, по переданным параметрам
        """
        all_books: list[Book | None] = self.books
        result: set[Book | None] = set()

        if title:
            result.update([book for book in all_books if title.lower() in book.title.lower()])
        if author:
            result.update([book for book in all_books if author.lower() in book.author.lower()])
        if year:
            result.update([book for book in all_books if book.year == year])

        return result if result else all_books


    def change_status(self, book: Book) -> None:
        """
        Метод меняющий состояние статуса на противоположный у книги, которая была заранее найдена
        """
        if book.status == BookStatus.IN_STOCK:
            book.status = BookStatus.CHECKED_OUT
        else:
            book.status = BookStatus.IN_STOCK
        self.storage.save_books(books=self.books)
