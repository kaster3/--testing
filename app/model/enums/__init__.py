__all__ = (
    "TaskStatus",
    "Category",
    "TaskPriority",
)

from .category import Category
from .priority import TaskPriority
from .status import TaskStatus
