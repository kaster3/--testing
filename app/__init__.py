__all__ = (
    "Storage",
    "Book",
    "BookStatus",
    "LibraryCRUD",
    "Service",
)

from .storage import Storage
from .model import Book, BookStatus
from .library import LibraryCRUD
from .service_layer import Service