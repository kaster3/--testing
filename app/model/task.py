from datetime import datetime

from .enums import Category, TaskPriority, TaskStatus


# @id_generator
class Task:
    """Основная бизнес сущность с которой будут выполняться все CRUD операции"""

    __slots__ = (
        "id",
        "title",
        "description",
        "category",
        "due_date",
        "priority",
        "status",
    )

    def __init__(
        self, id, title, description, category, due_date, priority, status=TaskStatus.INCOMPLETE
    ) -> None:
        self.id = id
        self.title: str = title
        self.description: str = description
        self.category: Category = category
        self.due_date: datetime = due_date
        self.priority: TaskPriority = priority
        self.status: TaskStatus = status

    def __str__(self):
        """
        Метод для вывода информации о задаче в удобном виде
        """
        return (
            f"Задача №{self.id}\nНазвание: {self.title}\nОписание:  {self.description}\n"
            f"Категория: {self.category.value}\nDeadline: {self.due_date}\n"
            f"Приоритет: {self.priority.value}\nСтатус выдачи: {self.status.value}\n"
        )

    def to_dict(self):
        """
        Метод для преобразования экземпляра задачи в словарь так как json не умеет
        работать с классом Enum и datetime по умолчанию
        """
        return {
            "id": self.id,
            "title": self.title,
            "description": self.description,
            "category": self.category.value,
            "due_date": self.due_date.strftime("%Y-%m-%d"),
            "priority": self.priority.value,
            "status": self.status.value,
        }
