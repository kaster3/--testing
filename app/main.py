from app.messages.main.consts import MAIN_EXIT, WRONG_INPUT
from app.messages.main.message_formatter import main_exception
from app.data_accesor_layer.repository import TaskRepository
from app.service_layer import TaskService
from app.service_layer.service_helper import ServiceHelper
from app.storage_layer.storage import TaskStorage


def main(service: TaskService) -> None:
    """
    Главный цикл приложения, который обрабатывает взаимодействие с пользователем и вызывает
    соответствующие методы у класса TaskService, выполняя основную бизнес логику
    """

    choices = {
        "1": service.show_tasks,
        "2": service.add_task,
        "3": service.update_task,
        "4": service.delete_task,
        "5": service.search_by_key_word,
        "6": lambda: print(MAIN_EXIT),
    }

    while True:
        choice = service.get_choice()
        choices.get(choice, lambda: print(WRONG_INPUT))()
        if choice == "6":
            break


if __name__ == "__main__":
    storage = TaskStorage()
    repository = TaskRepository(storage=storage)
    service_helper = ServiceHelper()
    service = TaskService(repository=repository, service_helper=service_helper)
    try:
        main(service=service)
    except Exception as exc:
        print(main_exception(exc=exc))
