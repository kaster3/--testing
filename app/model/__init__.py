__all__ = (
    "Task",
    "TaskPriority",
    "TaskStatus",
    "Category",
)

from .enums import Category, TaskPriority, TaskStatus
from .task import Task
