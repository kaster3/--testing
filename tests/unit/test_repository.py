from datetime import datetime

import pytest
from syrupy import SnapshotAssertion, snapshot

from app.model import Category, TaskPriority, TaskStatus


class TestTaskRepository:

    def test_get_all_tasks(self, repository_with_tasks, snapshot: SnapshotAssertion):
        all_tasks = repository_with_tasks.get_all_tasks()
        assert len(all_tasks) == 4
        snapshot.assert_match([task.to_dict() for task in all_tasks])

    def test_get_tasks_by_category(self, repository_with_tasks):
        category_tasks = repository_with_tasks.get_tasks_by_category(category=Category.JOB)
        assert len(category_tasks) == 2

    @pytest.mark.parametrize(
        "status, result",
        [
            (TaskStatus.INCOMPLETE, 3),
            (TaskStatus.COMPLETED, 1),
        ],
    )
    def test_get_tasks_by_status(self, status, result, repository_with_tasks, snapshot: SnapshotAssertion):
        tasks = repository_with_tasks.get_tasks_by_status(status=status)
        assert len(tasks) == result
        snapshot.assert_match([task.to_dict() for task in tasks])

    def test_add_task(self, repository, mock_storage, snapshot: SnapshotAssertion):
        task_data_1 = {
            "id": 5,
            "title": "New Task",
            "description": "This is a new task",
            "category": Category.PERSONAL,
            "due_date": datetime.strptime("2025-12-23", "%Y-%m-%d"),
            "priority": TaskPriority.HIGH,
            "status": TaskStatus.INCOMPLETE,
        }
        task_data_2 = {
            "id": 6,
            "title": "Task 2",
            "description": "This is task 2",
            "category": Category.TRAINING,
            "due_date": datetime.strptime("2025-12-23", "%Y-%m-%d"),
            "priority": TaskPriority.LOW,
            "status": TaskStatus.INCOMPLETE,
        }
        repository.add_task(task_data=task_data_1)
        repository.add_task(task_data=task_data_2)

        all_tasks = repository.get_all_tasks()
        assert len(all_tasks) == 2
        snapshot.assert_match([task.to_dict() for task in all_tasks])
        assert mock_storage.save_task_json.call_count == 2

    @pytest.mark.parametrize("task_id", [1, 2, 3])
    def test_get_task_by_id(self, repository_with_tasks, task_id, snapshot):
        task = repository_with_tasks.get_task_by_id(task_id)
        snapshot.assert_match(task.to_dict())

    def test_delete_task(self, repository_with_tasks, snapshot):
        tasks = repository_with_tasks.get_all_tasks()
        assert len(tasks) == 4
        repository_with_tasks.delete_task(task=tasks[0])
        assert len(tasks) == 3
        snapshot.assert_match([task.to_dict() for task in tasks])

    def test_update_task(self, repository_with_tasks, snapshot, mock_storage):
        task = repository_with_tasks.get_task_by_id(task_id=1)
        updated_data = {
            "title": "Updated Task",
            "description": "This is an updated task",
            "category": Category.PERSONAL,
            "due_date": datetime.strptime("2025-12-23", "%Y-%m-%d"),
            "priority": TaskPriority.HIGH,
            "status": TaskStatus.INCOMPLETE,
        }
        repository_with_tasks.update_task(task=task, updated_data=updated_data)
        updated_task = repository_with_tasks.get_task_by_id(task_id=1)
        snapshot.assert_match(updated_task.to_dict())
        assert mock_storage.save_task_json.call_count == 5

    @pytest.mark.parametrize(
        "key_word, result",
        [
            ("Task", 4),
            ("Высокий", 2),
            ("Описание", 1),
            ("Работа", 2),
            ("Не выполнена", 3),
            ("Выполнено", 1),
        ],
    )
    def test_search_by_key_word(self, key_word, result, repository_with_tasks, snapshot):
        tasks = repository_with_tasks.search_by_key_word(key_word=key_word)
        assert len(tasks) == result
        snapshot.assert_match([task.to_dict() for task in tasks])
