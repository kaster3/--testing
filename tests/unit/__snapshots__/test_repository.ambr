# serializer version: 1
# name: TestTaskRepository.test_add_task
  list([
    dict({
      'category': 'Личное',
      'description': 'This is a new task',
      'due_date': '2025-12-23',
      'id': 5,
      'priority': 'Высокий',
      'status': 'Не выполнена',
      'title': 'New Task',
    }),
    dict({
      'category': 'Обучение',
      'description': 'This is task 2',
      'due_date': '2025-12-23',
      'id': 6,
      'priority': 'Низкий',
      'status': 'Не выполнена',
      'title': 'Task 2',
    }),
  ])
# ---
# name: TestTaskRepository.test_delete_task
  list([
    dict({
      'category': 'Личное',
      'description': 'Description 2',
      'due_date': '2025-12-23',
      'id': 2,
      'priority': 'Низкий',
      'status': 'Не выполнена',
      'title': 'Task 2',
    }),
    dict({
      'category': 'Обучение',
      'description': '',
      'due_date': '2025-12-23',
      'id': 3,
      'priority': 'Низкий',
      'status': 'Выполнено',
      'title': 'Task 3',
    }),
    dict({
      'category': 'Работа',
      'description': 'Description 4',
      'due_date': '2025-12-23',
      'id': 4,
      'priority': 'Высокий',
      'status': 'Не выполнена',
      'title': 'Task 4',
    }),
  ])
# ---
# name: TestTaskRepository.test_get_all_tasks
  list([
    dict({
      'category': 'Работа',
      'description': 'Описание 1',
      'due_date': '2025-12-23',
      'id': 1,
      'priority': 'Высокий',
      'status': 'Не выполнена',
      'title': 'Task 1',
    }),
    dict({
      'category': 'Личное',
      'description': 'Description 2',
      'due_date': '2025-12-23',
      'id': 2,
      'priority': 'Низкий',
      'status': 'Не выполнена',
      'title': 'Task 2',
    }),
    dict({
      'category': 'Обучение',
      'description': '',
      'due_date': '2025-12-23',
      'id': 3,
      'priority': 'Низкий',
      'status': 'Выполнено',
      'title': 'Task 3',
    }),
    dict({
      'category': 'Работа',
      'description': 'Description 4',
      'due_date': '2025-12-23',
      'id': 4,
      'priority': 'Высокий',
      'status': 'Не выполнена',
      'title': 'Task 4',
    }),
  ])
# ---
# name: TestTaskRepository.test_get_task_by_id[1]
  dict({
    'category': 'Работа',
    'description': 'Описание 1',
    'due_date': '2025-12-23',
    'id': 1,
    'priority': 'Высокий',
    'status': 'Не выполнена',
    'title': 'Task 1',
  })
# ---
# name: TestTaskRepository.test_get_task_by_id[2]
  dict({
    'category': 'Личное',
    'description': 'Description 2',
    'due_date': '2025-12-23',
    'id': 2,
    'priority': 'Низкий',
    'status': 'Не выполнена',
    'title': 'Task 2',
  })
# ---
# name: TestTaskRepository.test_get_task_by_id[3]
  dict({
    'category': 'Обучение',
    'description': '',
    'due_date': '2025-12-23',
    'id': 3,
    'priority': 'Низкий',
    'status': 'Выполнено',
    'title': 'Task 3',
  })
# ---
# name: TestTaskRepository.test_get_tasks_by_status[TaskStatus.COMPLETED-1]
  list([
    dict({
      'category': 'Обучение',
      'description': '',
      'due_date': '2025-12-23',
      'id': 3,
      'priority': 'Низкий',
      'status': 'Выполнено',
      'title': 'Task 3',
    }),
  ])
# ---
# name: TestTaskRepository.test_get_tasks_by_status[TaskStatus.INCOMPLETE-3]
  list([
    dict({
      'category': 'Работа',
      'description': 'Описание 1',
      'due_date': '2025-12-23',
      'id': 1,
      'priority': 'Высокий',
      'status': 'Не выполнена',
      'title': 'Task 1',
    }),
    dict({
      'category': 'Личное',
      'description': 'Description 2',
      'due_date': '2025-12-23',
      'id': 2,
      'priority': 'Низкий',
      'status': 'Не выполнена',
      'title': 'Task 2',
    }),
    dict({
      'category': 'Работа',
      'description': 'Description 4',
      'due_date': '2025-12-23',
      'id': 4,
      'priority': 'Высокий',
      'status': 'Не выполнена',
      'title': 'Task 4',
    }),
  ])
# ---
# name: TestTaskRepository.test_search_by_key_word[Task-4]
  list([
    dict({
      'category': 'Работа',
      'description': 'Описание 1',
      'due_date': '2025-12-23',
      'id': 1,
      'priority': 'Высокий',
      'status': 'Не выполнена',
      'title': 'Task 1',
    }),
    dict({
      'category': 'Личное',
      'description': 'Description 2',
      'due_date': '2025-12-23',
      'id': 2,
      'priority': 'Низкий',
      'status': 'Не выполнена',
      'title': 'Task 2',
    }),
    dict({
      'category': 'Обучение',
      'description': '',
      'due_date': '2025-12-23',
      'id': 3,
      'priority': 'Низкий',
      'status': 'Выполнено',
      'title': 'Task 3',
    }),
    dict({
      'category': 'Работа',
      'description': 'Description 4',
      'due_date': '2025-12-23',
      'id': 4,
      'priority': 'Высокий',
      'status': 'Не выполнена',
      'title': 'Task 4',
    }),
  ])
# ---
# name: TestTaskRepository.test_search_by_key_word[\u0412\u044b\u043f\u043e\u043b\u043d\u0435\u043d\u043e-1]
  list([
    dict({
      'category': 'Обучение',
      'description': '',
      'due_date': '2025-12-23',
      'id': 3,
      'priority': 'Низкий',
      'status': 'Выполнено',
      'title': 'Task 3',
    }),
  ])
# ---
# name: TestTaskRepository.test_search_by_key_word[\u0412\u044b\u0441\u043e\u043a\u0438\u0439-2]
  list([
    dict({
      'category': 'Работа',
      'description': 'Описание 1',
      'due_date': '2025-12-23',
      'id': 1,
      'priority': 'Высокий',
      'status': 'Не выполнена',
      'title': 'Task 1',
    }),
    dict({
      'category': 'Работа',
      'description': 'Description 4',
      'due_date': '2025-12-23',
      'id': 4,
      'priority': 'Высокий',
      'status': 'Не выполнена',
      'title': 'Task 4',
    }),
  ])
# ---
# name: TestTaskRepository.test_search_by_key_word[\u041d\u0435 \u0432\u044b\u043f\u043e\u043b\u043d\u0435\u043d\u0430-3]
  list([
    dict({
      'category': 'Работа',
      'description': 'Описание 1',
      'due_date': '2025-12-23',
      'id': 1,
      'priority': 'Высокий',
      'status': 'Не выполнена',
      'title': 'Task 1',
    }),
    dict({
      'category': 'Личное',
      'description': 'Description 2',
      'due_date': '2025-12-23',
      'id': 2,
      'priority': 'Низкий',
      'status': 'Не выполнена',
      'title': 'Task 2',
    }),
    dict({
      'category': 'Работа',
      'description': 'Description 4',
      'due_date': '2025-12-23',
      'id': 4,
      'priority': 'Высокий',
      'status': 'Не выполнена',
      'title': 'Task 4',
    }),
  ])
# ---
# name: TestTaskRepository.test_search_by_key_word[\u041e\u043f\u0438\u0441\u0430\u043d\u0438\u0435-1]
  list([
    dict({
      'category': 'Работа',
      'description': 'Описание 1',
      'due_date': '2025-12-23',
      'id': 1,
      'priority': 'Высокий',
      'status': 'Не выполнена',
      'title': 'Task 1',
    }),
  ])
# ---
# name: TestTaskRepository.test_search_by_key_word[\u0420\u0430\u0431\u043e\u0442\u0430-2]
  list([
    dict({
      'category': 'Работа',
      'description': 'Описание 1',
      'due_date': '2025-12-23',
      'id': 1,
      'priority': 'Высокий',
      'status': 'Не выполнена',
      'title': 'Task 1',
    }),
    dict({
      'category': 'Работа',
      'description': 'Description 4',
      'due_date': '2025-12-23',
      'id': 4,
      'priority': 'Высокий',
      'status': 'Не выполнена',
      'title': 'Task 4',
    }),
  ])
# ---
# name: TestTaskRepository.test_update_task
  dict({
    'category': 'Личное',
    'description': 'This is an updated task',
    'due_date': '2025-12-23',
    'id': 1,
    'priority': 'Высокий',
    'status': 'Не выполнена',
    'title': 'Updated Task',
  })
# ---
